### PR TITLE
Handle empty SERVER_PORT information on signature checks

### PR DESCRIPTION
**Pull request by leogermani**
Related to this JPOP issue 6089-gh-jpop-issues

### The bug

We are not handling situations where PHP variable `$_SERVER['SERVER_PORT']` is not set. We assume it is always there, but according to the PHP documentation](https://www.php.net/manual/en/reserved.variables.server.php):

```
Note: Under the Apache 2, you must set UseCanonicalName = On, as well as UseCanonicalPhysicalPort = On in order to get the physical (real) port, otherwise, this value can be spoofed and it may or may not return the physical port value. It is not safe to rely on this value in security-dependent contexts. 
```

When this value is not set, we cast it to integer and it ends up as `0` -> https://github.com/Automattic/jetpack/blob/master/packages/connection/legacy/class-jetpack-signature.php#L78

Then, when we check it further, `0` is not in the list of the default ports and it ends up in the URL, making it invalid. -> https://github.com/Automattic/jetpack/blob/master/packages/connection/legacy/class-jetpack-signature.php#L95

### Why

[16 Sep 2019](https://github.com/Automattic/jetpack/pull/13489) - We made the `in_array` check strict, to match our PHPCS rules

[24 Sep 2019](https://github.com/Automattic/jetpack/pull/13523) - This broke the check for the PORT number, because it is often a string, so the typecast was introduced

[22 Nov 2019](https://github.com/Automattic/jetpack/pull/14111) - The loose comparison was reintroduced, because it was breaking other situations

So the typecast was introduced to fix something that was removed later on and could safely be removed.

### The Fix

We could simply remove the line that does the typecast, but that would keep throwing an Error Notice when `$_SERVER['SERVER_PORT']` is not set, so we fixed that as well.

#### Jetpack product discussion

N/A

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Start with a connected site on the master branch
* Edit your `wp-config.php` and add `unset($_SERVER['SERVER_PORT']);`.
* Access the Jetpack debugger and debug your site
* Check that the debugger returns an error
* Check your local error logs and confirm it's throwing a PHP Error notice for the undefined variable

* Check out this branch
* Visit the debugger again
* Confirm Everything Looks Great
* Confirm there are no errors in your local logs

* Edit your `wp-config.php` and add `$_SERVER['HTTP_X_FORWARDED_PORT'] = 80`
* run the tests and check that everything works

#### Proposed changelog entry for your changes:

* Handle XMLRPC requests when SERVER_PORT is not defined

### CODE REVIEW

File: packages/connection/legacy/class-jetpack-signature.php [hunk 1]
```diff
@@ -74,8 +74,8 @@ public function sign_current_request( $override = array() ) {
 			}
 		}
 
-		$host_port = isset( $_SERVER['HTTP_X_FORWARDED_PORT'] ) ? $_SERVER['HTTP_X_FORWARDED_PORT'] : $_SERVER['SERVER_PORT'];
-		$host_port = (int) $host_port;
+		$host_port = isset( $_SERVER['HTTP_X_FORWARDED_PORT'] ) ? $_SERVER['HTTP_X_FORWARDED_PORT'] : '';
+		$host_port = '' === $host_port && isset( $_SERVER['SERVER_PORT'] ) ? $_SERVER['SERVER_PORT'] : $host_port;
```
Comment (anomiex) [inline packages/connection/legacy/class-jetpack-signature.php:78]: Would it be clearer to have an `if` here instead?

**Review by anomiex — COMMENTED** [review]
Code itself seems ok. You can ignore the inline comment if you want.

But, can we add tests for the failure cases here?

**Comment by leogermani** [comment]
@anomiex 

I experimented with the `@runInSeparateProcess` annotation.

It won't work if we don't disable the `@preserveGlobalState`. However, this won't work well with WorDBless, which is loaded in `bootstrap.php` and required by many tests. When WorDBless loads WordPress' core, it fails because some globals are missing.

Do you see any workaround? I can only think of refactoring the method (add a filter to the constants so we can filter the values in the tests) or try to refactor all the tests and only load WorDBless when needed, although I'm not sure this is going to work.

**Comment by anomiex** [comment]
Hmm. Poking at it a little bit, it doesn't seem that WorDBless itself is the problem. The problem is that `$_SERVER['SCRIPT_FILENAME']` winds up being empty, and WordPress's `wp_guess_url()` chokes on that.

<details>
<summary>If we set that to something reasonable in bootstrap.php, it seems to work.</summary>

```diff
diff --git a/packages/connection/tests/php/bootstrap.php b/packages/connection/tests/php/bootstrap.php
index cf4fe7303..cb6afe95e 100644
--- a/packages/connection/tests/php/bootstrap.php
+++ b/packages/connection/tests/php/bootstrap.php
@@ -10,4 +10,9 @@
  */
 require_once __DIR__ . '/../../vendor/autoload.php';
 
+// Work around WordPress bug when `@runInSeparateProcess` is used.
+if ( empty( $_SERVER['SCRIPT_FILENAME'] ) ) {
+       $_SERVER['SCRIPT_FILENAME'] = '/usr/local/bin/phpunit'; // arbitrary.
+}
+
 \WorDBless\Load::load();
```
</details>

<details>
<summary>(The WordPress bug itself, BTW)</summary>

```diff
diff --git a/src/wp-includes/functions.php b/src/wp-includes/functions.php
index 537425d898..20139e1763 100644
--- a/src/wp-includes/functions.php
+++ b/src/wp-includes/functions.php
@@ -5491,7 +5491,7 @@ function wp_guess_url() {
                                $directory = str_replace( ABSPATH, '', $script_filename_dir );
                                // Strip off the subdirectory, and any file/query params.
                                $path = preg_replace( '#/' . preg_quote( $directory, '#' ) . '/[^/]*$#i', '', $_SERVER['REQUEST_URI'] );
-                       } elseif ( false !== strpos( $abspath_fix, $script_filename_dir ) ) {
+                       } elseif ( '' !== $script_filename_dir && false !== strpos( $abspath_fix, $script_filename_dir ) ) {
                                // Request is hitting a file above ABSPATH.
                                $subdirectory = substr( $abspath_fix, strpos( $abspath_fix, $script_filename_dir ) + strlen( $script_filename_dir ) );
                                // Strip off any file/query params from the path, appending the subdirectory to the installation.
```
</details>

**Comment by leogermani** [comment]
Did you test it all the way through? I confess that when I got to this point it I just assumed WordPress would require a lot of other globals and this would be undoable. Guess I was too pessimistic.

**Comment by anomiex** [comment]
I tested it by adding a dummy test to the class, along the lines of
```
/**
 * @runInSeparateProcess
 * @preserveGlobalState disabled
 */
public function test_foo() {
    $this->assertTrue( true );
}
```
and then poked at it until I got it to pass.

**Comment by kraftbj** [comment]
@anomiex Could you summarize the WP bug and report it on https://core.trac.wordpress.org? 

I did a quick glance searching for existing tickets, but I'm not sure the best way to describe it so may be missing it.

File: packages/connection/legacy/class-jetpack-signature.php [hunk 2]
```diff
@@ -361,17 +365,40 @@ public function get_current_request_port() {
...
+	public function sanitize_host_post( $port_number ) {
+
+		if ( ! is_int( $port_number ) && ! is_string( $port_number ) ) {
+			return '';
+		}
+		if ( is_string( $port_number ) && ! ctype_digit( $port_number ) ) {
+			return '';
+		}
+
+		if ( 0 >= (int) $port_number ) {
```
Comment (anomiex) [inline packages/connection/legacy/class-jetpack-signature.php:399]: Eew, that's confusing. Yoda conditionals for the lose.

Comment (anomiex) [inline packages/connection/legacy/class-jetpack-signature.php:399]: Should you also check that the port isn't above 65535?

Comment (leogermani) [reply packages/connection/legacy/class-jetpack-signature.php:399]: good call. 

I tested the constants in many ways, but these have a big impact on the time taken to run tests. We have to consider when this is going to be a problem

Comment (leogermani) [reply packages/connection/legacy/class-jetpack-signature.php]: PR updated with the port range validation

Comment (anomiex) [reply packages/connection/legacy/class-jetpack-signature.php]: If we test the sanitization function that's used for the constants (which you are), there shouldn't be a need to test all failure cases for the constants themselves.

We could also look at where the overhead is actually coming from, and see if we can reduce it. Maybe your idea of not initializing WorDBless unnecessarily in bootstrap.php would be good for that purpose after all. But perhaps that can wait for a followup PR.

**Review by anomiex — APPROVED** [review]
Looks sane, and has lots of tests.

**Comment by anomiex** [comment]
> @anomiex Could you summarize the WP bug and report it on https://core.trac.wordpress.org (and bonus points for attaching the patch)?

Done, https://core.trac.wordpress.org/ticket/51773

File: packages/connection/legacy/class-jetpack-signature.php (hunk 2 above)
Comment (leogermani) [reply packages/connection/legacy/class-jetpack-signature.php]: that's what I was thinking. The method uses one WordPress function (`is_ssl`) that could be easily mocked.

But I'm not sure the impact would be too big. It seems to me that what takes time is to setup happens before we get to `bootstrap`, so not loading the WordPress files wouldn't change the time that much. But I can give it a try.

And I agree that only one test for the constants should be enough

Comment (leogermani) [reply packages/connection/legacy/class-jetpack-signature.php]: Confirmed that the time difference if we don't load WordPress files is irrelevant.

Removed 2 redundant constants tests

File: packages/connection/tests/php/test_Signature.php [hunk 3]
```diff
@@ -395,61 +395,19 @@ public function get_request_port_data_provider() {
...
-
-	/**
-	 * Runs isolated tests to check the behavior of constants with invalid values
-	 *
-	 * Uses @see self::test_get_request_port
-	 *
-	 * @runInSeparateProcess
-	 * @preserveGlobalState disabled
-	 */
-	public function test_request_port_constants_invalid_valies() {
```
Comment (anomiex) [inline packages/connection/tests/php/test_Signature.php]: We might still want to keep a failure-case test.

**Review by anomiex — APPROVED** [review]

**Comment by anomiex** [comment]
FYI, I see this has been tagged as "ready to merge" for a few days. The code freeze for 9.2 is today, if you want to get this into that release you'll need to get it merged in the next few hours. If you'd rather wait for 9.3, that's fine too, just change the milestone or let me know so I can change it for you.

**Comment by leogermani** [comment]
> FYI, I see this has been tagged as "ready to merge" for a few days. The code freeze for 9.2 is today, if you want to get this into that release you'll need to get it merged in the next few hours. If you'd rather wait for 9.3, that's fine too, just change the milestone or let me know so I can change it for you.

Thanks for the reminder! I think this is ready to land, but the tests are failing, Let me have a look
